rust-symcrypt/src/rsa/mod.rs

```diff
@@ -0,0 +1,755 @@
+            let prime_2 = symcrypt_sys::SymCryptRsakeySizeofPrime(self.inner.0, 1);
+            (prime_1, prime_2)
+        }
+    }
+
+    /// `get_size_of_modulus()` returns a `u32` representing the (tight) size in bytes of a byte array big enough to store
+    /// the modulus of the key.
+    pub fn get_size_of_modulus(&self) -> u32 {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeySizeofModulus(self.inner.0)
+        }
+    }
+
+    /// `get_size_of_public_exponent()` returns a `u32` representing the (tight) size in bytes of a byte array big enough to store
+    /// the public exponent of the key.
+    pub fn get_size_of_public_exponent(&self) -> u32 {
+        unsafe {
+            // SAFETY: FFI calls
+            // Only one public exponent is supported, so the only valid index is 0.
+            symcrypt_sys::SymCryptRsakeySizeofPublicExponent(self.inner.0, 0)
+        }
+    }
+
+    /// `get_rsa_key_usage()` returns the intended usage for the RSA key pair.
+    pub fn get_rsa_key_usage(&self) -> RsaKeyUsage {
+        self.rsa_key_usage
+    }
+
+    // `inner()` gives crate access to the inner symcrypt Rsa Key object.
+    pub(crate) fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY {
+        self.inner.0
+    }
+}
+
+// Utility function to reduce common RSA allocation call
+fn allocate_rsa(
+    n_primes: u32,
+    n_bits_mod: symcrypt_sys::SIZE_T,
+) -> Result<InnerRsaKey, SymCryptError> {
+    let rsa_params = symcrypt_sys::SYMCRYPT_RSA_PARAMS {
+        version: 1 as symcrypt_sys::UINT32, // No other version aside from version 1 is specified
+        nBitsOfModulus: n_bits_mod as symcrypt_sys::UINT32,
+        nPrimes: n_primes as symcrypt_sys::UINT32, // To generate a only a public key, 0 must be passed.
+        nPubExp: 1 as symcrypt_sys::UINT32,        // Value must be 1.
+    };
+    unsafe {
+        // SAFETY: FFI calls
+        let result = symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0);
+        if result == ptr::null_mut() {
+            return Err(SymCryptError::AuthenticationFailure);
+        }
+        Ok(InnerRsaKey(result))
+    }
+}
+
+// Utility function to store a `u64` into a new byte vector in big-endian format.
+fn store_msb_first_u64(value: u64, size: u32) -> Result<Vec<u8>, SymCryptError> {
+    let mut dst = vec![0u8; size as usize]; // Allocate tight size in bytes for storing public exponent
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptStoreMsbFirstUint64(value, dst.as_mut_ptr(), size as u64) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(dst),
+            err => Err(SymCryptError::from(err)),
+        }
+    }
+}
+
+/// Load a `u64` from a byte slice assuming big-endian order.
+fn load_msb_first_u64(src: &[u8]) -> Result<u64, SymCryptError> {
+    let mut dst = 0u64;
+    unsafe {
+        // SAFETY: FFI calls
+        match symcrypt_sys::SymCryptLoadMsbFirstUint64(
+            src.as_ptr(),
```

Comment:
>src [](http://example.com/codeflow?start=12&length=3)

I'm assuming this can have leading 0 bytes. Also, doesn't need to be length of 8 bytes?